design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,483 @@
+    This is not part of a user-facing API and will not be read by the operator to inform any
+    operations. It's only purpose is to guarantee pod restart any time the key is rotated.
+8. IFF keys were created/rotated, Rook updates the appropriate cephx key status(es)
+    1. `status...keyGeneration` is taken to be `spec...keyGeneration`
+    2. `status...keyCephVersion` is taken to be `cephImageVersion`
+
+### CephCluster example
+
+1. CephCluster daemons are all daemon, so the reconcile reads
+    `spec.security.cephx.daemon` to determine if the reconcile should rotate.
+2. When key rotation is indicated:
+    1. Rotate the admin key. This is the first key created by Rook and will be the first rotated.
+        This should be as easy as rotating it and updating the daemon reference. However, depending
+        on how child reconciliations are updated with the admin key, it may be necessary to restart
+        the Rook operator reconcile manager (all reconcile routines) after this rotation.
+    2. Update the CephCluster status with the admin key info
+    3. Rotate the mon key. Mon key rotation requires a special process:
+        1. Update mon deployments first without rotating the key
+        2. Rotate the key
+        3. Restart the mon deployments to cause them to get the rotated key
+    4. Update the CephCluster status with the mon key info
+    5. For each mgr, rotate the key and update/restart the mgr daemon
+    6. Update the CephCluster status with the mgr key info
+    7. For each OSD, rotate the key
+        1. OSDs don't have an associated keyring secret, so we must make sure all OSD deployments'
+            "activate" init container gets the latest key and updates it for the "osd" pod
+    8. Update the CephCluster status with the OSD key info
+
+With separated daemon key info tracking, the status will look like so:
+
+```yaml
+status:
+  # ...
+  cephx:
+    admin:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mon:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    mgr:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    osd:
+      keyGeneration: 3 # e.g.
+      keyCephVersion: "20.2.2" # e.g.
+    csi: {} # discussed more below, but some CSI keys and need to be in this status
+```
+
+Keeping track of each daemon type separately helps Rook ensure it won't re-rotate keys from earlier
+in the reconcile if the reconcile needs to restart later in the process. For example, there is no
+need to re-rotate admin/mon/mgr keys in the event the reconcile needs to restart in the middle of
+OSD updates. Because OSD updates can take quite a while, this case is likely to occur across Rook's
+large user base.
+
+### CSI example
+
+CSI keys are updated as part of a CephCluster reconcile.
+
+For a CephCluster reconcile:
+
+1. Ceph-CSI daemon keys (provisioners and node plugins) use `CephCluster.spec.security.cephx.csi`
```

Comment:
If I understand correctly, for csi rotation will happen  as a overlapping rotation, During ceph upgrades

Which means once all the csi keys are rotated, we need to reboot/drain the nodes?
But that is not a part of upgrade process for minor releases, so does user needs to manually do reboot nodes, for the older pvcs that are mounted

Reply:
That only comes into play if the `WithCephVersionUpdate` is used for CSI. That option will likely be disabled for CSI so users won't unexpectedly lose connections to PVCs.